<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="decorate">
<title>decorate: Differential Epigenetic Correlation Test â€¢ decorate</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="decorate: Differential Epigenetic Correlation Test">
<meta property="og:description" content="decorate">
<meta property="og:image" content="http://gabrielhoffman.github.io/decorrelate/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">decorate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.30</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/decorate.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/decorate/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>decorate: Differential Epigenetic Correlation Test</h1>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2022-09-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/decorate/blob/HEAD/vignettes/decorate.Rmd" class="external-link"><code>vignettes/decorate.Rmd</code></a></small>
      <div class="d-none name"><code>decorate.Rmd</code></div>
    </div>

    
    
<!--- 
# # run analysis
# cd /Users/gabrielhoffman/workspace/repos/decorate/vignettes
# # cd /hpc/users/hoffmg01/work/pinnacle/vignettes
# # cd ~/work//decorate/
# rmarkdown::render("decorate_example.Rmd", output_dir='./', intermediates_dir='./'); system("cat decorate_example.html | grep -v 'Found more than one class' | grep -v 'PythonEmbedInR' > decorate_example2.html") 

# scp decorate_example2.html sklar1:/sc/orga/work/hoffmg01/software/decorate/decorate_example.html 

# cd /Users/gabrielhoffman/workspace/scripts/decorate_analysis
# \cp -f EpiMap.Rmd /Users/gabrielhoffman/workspace/repos/decorate_analysis
# \cp -f KIRC.Rmd /Users/gabrielhoffman/workspace/repos/decorate_analysis
# \cp -f atac_local_corr.Rmd /Users/gabrielhoffman/workspace/repos/decorate_analysis
# \cp -f methyl_scz_decorate.Rmd /Users/gabrielhoffman/workspace/repos/decorate_analysis
# \cp -f simulations.Rmd /Users/gabrielhoffman/workspace/repos/decorate_analysis


# # move simulation to website on *minerva*
# cd /hpc/users/hoffmg01/www/software/decorate
# \cp -f /sc/orga/projects/psychencode/gabriel/decorate_analysis/KIRC2.html KIRC.html
# \cp -f /sc/orga/projects/psychencode/gabriel/decorate_analysis/EpiMap2.html EpiMap.html
# \cp -f /sc/orga/projects/psychencode/gabriel/decorate_analysis/atac_local_corr2.html atac_local_corr.html
# \cp -f /sc/orga/projects/psychencode/gabriel/decorate_analysis/methyl_scz_decorate2.html methyl_scz_decorate.html
# \cp -f /sc/orga/projects/psychencode/gabriel/decorate_analysis/simulations.html simulations.html


--->
<p>Standard analysis compares the differences in magnitude in epigenetic
signal between two subsets of a dataset. The decorate workflow evaluates
the local correlation structure between nearby epigenetic features and
identify clusters of features that are differential correlation between
two subsets of a dataset. Epigenetic datasets where decorate can be
applied include ChIP-seq, ATAC-seq and DNA methylation.</p>
<p>decorate v1.0.30</p>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<ol start="0" style="list-style-type: decimal">
<li>
<p>Data normalization and covariate correction</p>
<ul>
<li>We assume this has already been perform using standard methods for
your assay</li>
<li>decorate should be run on residuals after removing the effects of
confounding variables</li>
</ul>
</li>
</ol>
<ol style="list-style-type: decimal">
<li>
<p>Compute correlations between local pairs of features</p>
<ul>
<li><code><a href="../reference/runOrderedClusteringGenome.html">runOrderedClusteringGenome()</a></code></li>
</ul>
</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>
<p>Perform hierarchical clustering of features</p>
<ul>
<li>built into <code><a href="../reference/runOrderedClusteringGenome.html">runOrderedClusteringGenome()</a></code>
</li>
</ul>
</li>
</ol>
<ol start="3" style="list-style-type: decimal">
<li>
<p>Produce discrete clusters of features</p>
<ul>
<li>
<code>createClusters():</code> can take multiple parameter values to
return multiple sets of clusters at different resolutions</li>
</ul>
</li>
</ol>
<ol start="4" style="list-style-type: decimal">
<li>
<p>Filter clusters</p>
<ul>
<li>Based on strength of correlation
<ul>
<li>
<code>scoreClusters():</code> evaluate the strength of the
correlation structure of each cluster</li>
<li>
<code>retainClusters():</code> identify clusters that pass a cutoff
for strength of the correlation structure. By default filters by LEF,
the lead eigen-value fraction, which is the fraction of variance
explained by the first eigen-value. Higher LEF values indicate a
stronger correlation structure.</li>
<li>
<code>filterClusters():</code> apply filter to clusters based on
this cutoff</li>
</ul>
</li>
<li>Identify overlapping clusters and drop if redundant
<ul>
<li>
<code>collapseClusters():</code> Collapse redundant clusters
identified from using multiple parameter values in step (3). Redundancy
of two clusters is evaluated using the Jaccard index, the fraction of
epigenetic features shared between two clusters.</li>
</ul>
</li>
</ul>
</li>
</ol>
<ol start="5" style="list-style-type: decimal">
<li>
<p>Statistical test of differential correlation</p>
<ul>
<li>
<code>evalDiffCorr():</code> implements 16 tests of differential
correlation with different properties. See <a href="https://hoffmg01.u.hpc.mssm.edu/software/decorate/simulations.html" class="external-link">here</a>
for comparison of methods. I recommend â€˜Box.permuteâ€™, â€˜deltaSLEâ€™, â€˜sLEDâ€™
to retain power while controlling the false positive rate</li>
<li>merge all results into one data.frame with
<code><a href="../reference/combineResults.html">combineResults()</a></code>
</li>
</ul>
</li>
</ol>
<ol start="6" style="list-style-type: decimal">
<li>Data visualization
<ul>
<li>
<code>plotDecorate():</code> plot local correlation structure,
clusters of features and genes in the region</li>
<li>
<code>plotCompareCorr():</code> plot correlation structure for two
subsets of the data</li>
<li>
<code><a href="../reference/plotScatterPairs.html">plotScatterPairs()</a></code>: scatter plot for each pair of
features for two subsets of the data</li>
</ul>
</li>
</ol>
</div>
<div class="section level2">
<h2 id="analysis-of-simulated-data">Analysis of simulated data<a class="anchor" aria-label="anchor" href="#analysis-of-simulated-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DiseaseNeuroGenomics/decorate" class="external-link">decorate</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span>     </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span>     </span>
<span></span>
<span><span class="co"># load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'decorateData'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute residuals with respect to variable of interested plus confounders</span></span>
<span><span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">Disease</span>, <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">simData</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">residValues</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/residuals.html" class="external-link">residuals</a></span><span class="op">(</span> <span class="va">fit</span>, <span class="va">simData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Evaluate hierarchical clustering</span></span>
<span><span class="co"># adjacentCount is the number of adjacent peaks considered in correlation</span></span>
<span><span class="co"># use Spearman correlation to reduce the effects of outliers</span></span>
<span><span class="va">treeList</span> <span class="op">=</span> <span class="fu"><a href="../reference/runOrderedClusteringGenome.html">runOrderedClusteringGenome</a></span><span class="op">(</span> <span class="va">residValues</span>, <span class="va">simLocation</span>, adjacentCount<span class="op">=</span><span class="fl">500</span>, method.corr<span class="op">=</span><span class="st">"spearman"</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># Choose cutoffs and return clusters using multiple values for meanClusterSize </span></span>
<span><span class="co"># Clusters corresponding to each parameter value are returned </span></span>
<span><span class="co"># and then processed downstream</span></span>
<span><span class="co"># By using multiple parameter values, epigenetic features are included in clusters </span></span>
<span><span class="co"># at different resolutions</span></span>
<span><span class="va">treeListClusters</span> <span class="op">=</span> <span class="fu"><a href="../reference/createClusters.html">createClusters</a></span><span class="op">(</span> <span class="va">treeList</span>, method <span class="op">=</span> <span class="st">"meanClusterSize"</span>, meanClusterSize<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">40</span>, <span class="fl">50</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># Evaluate strength of correlation for each cluster</span></span>
<span><span class="va">clstScore</span> <span class="op">=</span> <span class="fu"><a href="../reference/scoreClusters.html">scoreClusters</a></span><span class="op">(</span><span class="va">treeList</span>, <span class="va">treeListClusters</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter to retain only strong clusters</span></span>
<span><span class="co"># If lead eigen value fraction (LEF) &gt; 30% then keep clusters</span></span>
<span><span class="co"># LEF is the fraction of variance explained by the first eigen-value</span></span>
<span><span class="va">clustInclude</span> <span class="op">=</span> <span class="fu"><a href="../reference/retainClusters.html">retainClusters</a></span><span class="op">(</span> <span class="va">clstScore</span>, <span class="st">"LEF"</span>, <span class="fl">0.15</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># get retained clusters</span></span>
<span><span class="va">treeListClusters_filter</span> <span class="op">=</span> <span class="fu"><a href="../reference/filterClusters.html">filterClusters</a></span><span class="op">(</span> <span class="va">treeListClusters</span>, <span class="va">clustInclude</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># collapse redundant clusters</span></span>
<span><span class="va">treeListClusters_collapse</span> <span class="op">=</span> <span class="fu"><a href="../reference/collapseClusters.html">collapseClusters</a></span><span class="op">(</span> <span class="va">treeListClusters_filter</span>, <span class="va">simLocation</span>, jaccardCutoff<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot correlations and clusters in region defind by query</span></span>
<span><span class="co"># get single entry giving range of the region</span></span>
<span><span class="va">query</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">simLocation</span><span class="op">)</span>  </span></code></pre></div>
<div class="section level4">
<h4 id="summary-of-clustering">Summary of clustering<a class="anchor" aria-label="anchor" href="#summary-of-clustering"></a>
</h4>
<p>Number of cluster parameters: 4</p>
<ul>
<li><code>meanClusterSize = c(20, 30, 40, 50)</code></li>
</ul>
<p>Number of total clusters: 57</p>
<ul>
<li>in <code>treeListClusters</code>
</li>
</ul>
<p>Number of clusters after filtering: 56</p>
<ul>
<li>in <code>treeListClusters_filter</code>
</li>
</ul>
<p>Number of clusters after collapsing: 29</p>
<ul>
<li>in <code>treeListClusters_collapse</code>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="make-plots">Make plots<a class="anchor" aria-label="anchor" href="#make-plots"></a>
</h3>
<div class="section level4">
<h4 id="evaluate-correlation-structure-versus-distance">Evaluate correlation structure versus distance<a class="anchor" aria-label="anchor" href="#evaluate-correlation-structure-versus-distance"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dfDist</span> <span class="op">=</span> <span class="fu"><a href="../reference/evaluateCorrDecay.html">evaluateCorrDecay</a></span><span class="op">(</span> <span class="va">treeList</span>, <span class="va">simLocation</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotCorrDecay.html">plotCorrDecay</a></span><span class="op">(</span> <span class="va">dfDist</span>, outlierQuantile<span class="op">=</span><span class="fl">1e-5</span> <span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/plot.corr.decay-1.png" width="672"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="plot-correlation-structure-along-genome">Plot correlation structure along genome<a class="anchor" aria-label="anchor" href="#plot-correlation-structure-along-genome"></a>
</h2>
<div class="section level4">
<h4 id="load-gene-database">Load gene database<a class="anchor" aria-label="anchor" href="#load-gene-database"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load gene locations</span></span>
<span><span class="co"># this is ENSEMBL v86 from the hg38 assembly</span></span>
<span><span class="co"># but other versions and assemblies are available</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: ensembldb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicFeatures</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationDbi</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationFilter</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'ensembldb'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ensembldb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicFeatures</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationDbi</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationFilter</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'ensembldb'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fig1</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotDecorate.html">plotDecorate</a></span><span class="op">(</span> <span class="va">ensdb</span>, <span class="va">treeList</span>, <span class="va">treeListClusters</span>, <span class="va">simLocation</span>, <span class="va">query</span><span class="op">)</span></span>
<span><span class="va">fig2</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotDecorate.html">plotDecorate</a></span><span class="op">(</span> <span class="va">ensdb</span>, <span class="va">treeList</span>, <span class="va">treeListClusters_filter</span>, <span class="va">simLocation</span>, <span class="va">query</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span> <span class="va">fig1</span>, <span class="va">fig2</span>, ncol<span class="op">=</span><span class="fl">2</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A: All clusters'</span>, <span class="st">'B: After filtering'</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/plot2Decorate-1.png" width="1440"></p>
</div>
<div class="section level3">
<h3 id="plot-after-filtering-and-collapsing-redundant-clusters">Plot after filtering and collapsing redundant clusters<a class="anchor" aria-label="anchor" href="#plot-after-filtering-and-collapsing-redundant-clusters"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fig</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotDecorate.html">plotDecorate</a></span><span class="op">(</span> <span class="va">ensdb</span>, <span class="va">treeList</span>, <span class="va">treeListClusters_collapse</span>, <span class="va">simLocation</span>, <span class="va">query</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span> <span class="va">fig</span>, ncol<span class="op">=</span><span class="fl">1</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'C: After collapsing clusters'</span><span class="op">)</span>, hjust<span class="op">=</span><span class="fl">0</span> <span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/plotCollapse.2-1.png" width="672"></p>
<div class="section level4">
<h4 id="run-test-of-differential-correlation-using-sled">Run test of differential correlation using sLED<a class="anchor" aria-label="anchor" href="#run-test-of-differential-correlation-using-sled"></a>
</h4>
<p>Evaluate <code>treeListClusters_collapse</code>, the filtered
collapsed set of clusters</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">4</span>, progressbar<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Evaluate Differential Correlation between two subsets of data</span></span>
<span><span class="co"># Use Spearman correlation to reduce the effect of outliers</span></span>
<span><span class="va">resDiffCorr</span> <span class="op">=</span> <span class="fu"><a href="../reference/evalDiffCorr-methods.html">evalDiffCorr</a></span><span class="op">(</span> <span class="va">residValues</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">Disease</span>, <span class="va">simLocation</span>, </span>
<span>  <span class="va">treeListClusters_collapse</span>, method<span class="op">=</span><span class="st">'deltaSLE'</span>,  method.corr<span class="op">=</span><span class="st">"spearman"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get summary of results</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span> <span class="va">resDiffCorr</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># print results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   id chrom cluster       pValue          stat n.perm     p.adjust</span></span>
<span><span class="co">## 1 40 chr20       2 3.984297e-05  0.0011542191      0 0.0009562313</span></span>
<span><span class="co">## 2 30 chr20       3 9.300193e-04  0.0007217765      0 0.0111602311</span></span>
<span><span class="co">## 3 20 chr20       4 6.301672e-03  0.0004177436      0 0.0504133770</span></span>
<span><span class="co">## 4 20 chr20       7 2.746754e-02  0.0004020404      0 0.1648052448</span></span>
<span><span class="co">## 5 40 chr20       9 7.185602e-02 -0.0001746903      0 0.3449089171</span></span>
<span><span class="co">## 6 20 chr20       6 9.529629e-02  0.0004449277      0 0.3811851459</span></span></code></pre>
<p>The id column gives the clustering parameter from
<code><a href="../reference/runOrderedClusteringGenome.html">runOrderedClusteringGenome()</a></code>, so cluster 6 with id 10 is
different from cluster 6 with id 50. <!---
Note that n.perm can vary across clusters because decorate uses a data-adaptive method to select the number of permutations.  Decorate devotes more permutations to clusters with smaller p-values 
---> Combine results to merge properties of each cluster into a single
data.frame</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_results</span> <span class="op">=</span> <span class="fu"><a href="../reference/combineResults.html">combineResults</a></span><span class="op">(</span> <span class="va">resDiffCorr</span>, <span class="va">clstScore</span>, <span class="va">treeListClusters</span>, <span class="va">simLocation</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df_results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   id chrom cluster       pValue          stat n.perm     p.adjust  N</span></span>
<span><span class="co">## 1 40 chr20       2 3.984297e-05  0.0011542191      0 0.0009562313 12</span></span>
<span><span class="co">## 2 30 chr20       3 9.300193e-04  0.0007217765      0 0.0111602311  9</span></span>
<span><span class="co">## 3 20 chr20       4 6.301672e-03  0.0004177436      0 0.0504133770  6</span></span>
<span><span class="co">## 4 20 chr20       7 2.746754e-02  0.0004020404      0 0.1648052448 26</span></span>
<span><span class="co">## 5 40 chr20       9 7.185602e-02 -0.0001746903      0 0.3449089171 13</span></span>
<span><span class="co">## 6 20 chr20       6 9.529629e-02  0.0004449277      0 0.3811851459 13</span></span>
<span><span class="co">##   mean_abs_corr quantile75 quantile90 quantile95       LEF    start      end</span></span>
<span><span class="co">## 1     0.1875826  0.2148852  0.2366822  0.2484625 0.1703609 62065457 62067875</span></span>
<span><span class="co">## 2     0.1843397  0.2106810  0.2366822  0.2473400 0.1941585 62066063 62067875</span></span>
<span><span class="co">## 3     0.2106171  0.2260056  0.2481257  0.2549663 0.2545741 62066417 62067325</span></span>
<span><span class="co">## 4     0.3702191  0.4085017  0.4338798  0.4433801 0.2175939 62071328 62078427</span></span>
<span><span class="co">## 5     0.3330685  0.3826729  0.4064634  0.4214374 0.2323639 62146651 62149161</span></span>
<span><span class="co">## 6     0.3092056  0.3520966  0.3790616  0.4096076 0.2208019 62068330 62071263</span></span>
<span><span class="co">##   width</span></span>
<span><span class="co">## 1  2419</span></span>
<span><span class="co">## 2  1813</span></span>
<span><span class="co">## 3   909</span></span>
<span><span class="co">## 4  7100</span></span>
<span><span class="co">## 5  2511</span></span>
<span><span class="co">## 6  2934</span></span></code></pre>
<p>Note that permutations are only used for test <code>'sLED'</code>, so
is set to zero for all others.</p>
</div>
<div class="section level4">
<h4 id="summary-of-cluster-properties">Summary of cluster properties<a class="anchor" aria-label="anchor" href="#summary-of-cluster-properties"></a>
</h4>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Histogram of LEF</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_results</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">LEF</span>, fill<span class="op">=</span><span class="va">id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">17</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"bottom"</span>, </span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html" class="external-link">scale_fill_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Requested mean cluster size"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Lead eigenvalue fraction (LEF)"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Summarize LEF"</span><span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/cluster%20properties-1.png" width="672"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Histogram of mean absolute correlation</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_results</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">mean_abs_corr</span>, fill<span class="op">=</span><span class="va">id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">17</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"bottom"</span>, </span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html" class="external-link">scale_fill_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Requested mean cluster size"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mean absolute correlation"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Summarize absolute correlation"</span><span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/cluster%20properties-2.png" width="672"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Boxplot of number of features per cluster</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_results</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">id</span>, <span class="va">N</span>, fill<span class="op">=</span><span class="va">id</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">17</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"bottom"</span>, </span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html" class="external-link">scale_fill_discrete</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Feature per cluster"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Requested mean cluster size"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Number of features"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Summarize feature per cluster"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/cluster%20properties-3.png" width="672"></p>
</div>
<div class="section level4">
<h4 id="compare-correlation-structure-in-two-subsets-of-the-data">Compare correlation structure in two subsets of the data<a class="anchor" aria-label="anchor" href="#compare-correlation-structure-in-two-subsets-of-the-data"></a>
</h4>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract feature identifiers from most significant cluster</span></span>
<span><span class="va">peakIDs</span> <span class="op">=</span> <span class="fu"><a href="../reference/getFeaturesInCluster.html">getFeaturesInCluster</a></span><span class="op">(</span> <span class="va">treeListClusters_collapse</span>, <span class="va">df_results</span><span class="op">$</span><span class="va">chrom</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">df_results</span><span class="op">$</span><span class="va">cluster</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">df_results</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">query</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">simLocation</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">simLocation</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">peakIDs</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">locText</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">seqnames</span>, <span class="st">':'</span>, <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">start</span>, big.mark<span class="op">=</span><span class="st">','</span><span class="op">)</span>, <span class="st">'-'</span>, <span class="fu"><a href="https://rdrr.io/r/base/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">end</span>, big.mark<span class="op">=</span><span class="st">','</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot comparison of correlation matrices for peaks in peakIDs</span></span>
<span><span class="co">#  where data is subset by metadata$Disease</span></span>
<span><span class="va">main</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">chrom</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">': cluster '</span>, <span class="va">df</span><span class="op">$</span><span class="va">cluster</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">'\n'</span>, <span class="va">locText</span><span class="op">)</span>   </span>
<span><span class="fu"><a href="../reference/plotCompareCorr.html">plotCompareCorr</a></span><span class="op">(</span> <span class="va">residValues</span>, <span class="va">peakIDs</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">Disease</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">main</span><span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/plotCompareCorr-1.png" width="384"></p>
</div>
<div class="section level4">
<h4 id="compare-correlation-structure-along-genome">Compare correlation structure along genome<a class="anchor" aria-label="anchor" href="#compare-correlation-structure-along-genome"></a>
</h4>
<p>First, plot correlation structure for controls (metadata$Disease==0).
Then, plot correlation structure for cases (metadata$Disease==1).</p>
<p>The cluster located at chr20:62,065,457-62,067,875 has a p-value of
3.984e-05 and a test statistic of 0.00115.<br><!---
The negative sign indicates a loss of correlation in the test set (i.e. Disease==1) compared to the baseline correlation in controls (i.e. Dissease==0)
---></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get location of peaks in this cluster  </span></span>
<span><span class="va">query</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">simLocation</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">simLocation</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">peakIDs</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># expand window to include adjacent clusters</span></span>
<span><span class="va">window</span> <span class="op">=</span> <span class="fl">2000</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span> <span class="op">-</span> <span class="va">window</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span> <span class="op">+</span> <span class="va">window</span></span>
<span></span>
<span><span class="va">fig1</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotDecorate.html">plotDecorate</a></span><span class="op">(</span> <span class="va">ensdb</span>, <span class="va">treeList</span>, <span class="va">treeListClusters_collapse</span>, <span class="va">simLocation</span>, <span class="va">query</span>, data<span class="op">=</span><span class="va">residValues</span><span class="op">[</span>,<span class="va">metadata</span><span class="op">$</span><span class="va">Disease</span><span class="op">==</span><span class="fl">0</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fig2</span> <span class="op">=</span> <span class="fu"><a href="../reference/plotDecorate.html">plotDecorate</a></span><span class="op">(</span> <span class="va">ensdb</span>, <span class="va">treeList</span>, <span class="va">treeListClusters_collapse</span>, <span class="va">simLocation</span>, <span class="va">query</span>, data<span class="op">=</span><span class="va">residValues</span><span class="op">[</span>,<span class="va">metadata</span><span class="op">$</span><span class="va">Disease</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span> <span class="va">fig1</span>, <span class="va">fig2</span>, ncol<span class="op">=</span><span class="fl">2</span>, labels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A: Contols (i.e. Disease==0)'</span>, <span class="st">'B: Cases (i.e. Disease==1)'</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/plot2Decorate.query-1.png" width="1440"></p>
</div>
<div class="section level4">
<h4 id="pairwise-scatter-plots">Pairwise scatter plots<a class="anchor" aria-label="anchor" href="#pairwise-scatter-plots"></a>
</h4>
<p>For all pairs of features in the significant cluster, make a
scatterplot for cases and controls seperately.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotScatterPairs.html">plotScatterPairs</a></span><span class="op">(</span> <span class="va">residValues</span>, <span class="va">peakIDs</span>, <span class="va">metadata</span><span class="op">$</span><span class="va">Disease</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">main</span><span class="op">)</span></span></code></pre></div>
<p><img src="decorate_files/figure-html/scatter-1.png" width="1440"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
