<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ decorate</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="http://gabrielhoffman.github.io/decorrelate/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">decorate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.31</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/decorate.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/decorate/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="boxM.html">boxM()</a></code> 
        </dt>
        <dd>Box's M-test</dd>
      </dl><dl><dt>
          
          <code><a href="boxM_fast.html">boxM_fast()</a></code> 
        </dt>
        <dd>Box's M-test</dd>
      </dl><dl><dt>
          
          <code><a href="boxM_permute.html">boxM_permute()</a></code> 
        </dt>
        <dd>Box's M-test</dd>
      </dl><dl><dt>
          
          <code><a href="collapseClusters.html">collapseClusters()</a></code> 
        </dt>
        <dd>Collapse clusters based on jaccard index</dd>
      </dl><dl><dt>
          
          <code><a href="combineResults.html">combineResults()</a></code> 
        </dt>
        <dd>Combine results into a single data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="corSubsetPairs.html">corSubsetPairs()</a></code> 
        </dt>
        <dd>Compute correlations between pairs of features</dd>
      </dl><dl><dt>
          
          <code><a href="corrMatrix.test.html">corrMatrix.test()</a></code> 
        </dt>
        <dd>Test difference between two correlation matricies</dd>
      </dl><dl><dt>
          
          <code><a href="countClusters-methods.html">countClusters()</a></code> 
        </dt>
        <dd>Count clusters on each chromosome</dd>
      </dl><dl><dt>
          
          <code><a href="createClusters.html">createClusters()</a></code> 
        </dt>
        <dd>Create cluster from list of hclust objects</dd>
      </dl><dl><dt>
          
          <code><a href="createCorrelationMatrix.html">createCorrelationMatrix()</a></code> 
        </dt>
        <dd>Create correlation matrix</dd>
      </dl><dl><dt>
          
          <code><a href="decorateData.html">decorateData</a></code> <code><a href="decorateData.html">decorateData</a></code> <code><a href="decorateData.html">decorateData</a></code> 
        </dt>
        <dd>Simulated data to show correlation clustering</dd>
      </dl><dl><dt>
          
          <code><a href="delaneau.score.html">delaneau.score()</a></code> 
        </dt>
        <dd>Score impact of each sample on correlation sturucture</dd>
      </dl><dl><dt>
          
          <code><a href="delaneau.test.html">delaneau.test()</a></code> 
        </dt>
        <dd>Test association between correlation sturucture and variable</dd>
      </dl><dl><dt>
          
          <code><a href="dot-evalDiffCorr.html">.evalDiffCorr()</a></code> 
        </dt>
        <dd>Internal .evalDiffCorr</dd>
      </dl><dl><dt>
          
          <code><a href="epiclust-class.html">epiclust-class</a></code> 
        </dt>
        <dd>Class epiclust</dd>
      </dl><dl><dt>
          
          <code><a href="epiclustDiscreteList-class.html">epiclustDiscreteList-class</a></code> 
        </dt>
        <dd>Class epiclustDiscreteList</dd>
      </dl><dl><dt>
          
          <code><a href="epiclustDiscreteListContain-class.html">epiclustDiscreteListContain-class</a></code> 
        </dt>
        <dd>Class epiclustDiscreteListContain</dd>
      </dl><dl><dt>
          
          <code><a href="epiclustList-class.html">epiclustList-class</a></code> 
        </dt>
        <dd>Class epiclustList</dd>
      </dl><dl><dt>
          
          <code><a href="evalDiffCorr-methods.html">evalDiffCorr()</a></code> 
        </dt>
        <dd>Evaluate Differential Correlation</dd>
      </dl><dl><dt>
          
          <code><a href="evaluateCorrDecay.html">evaluateCorrDecay()</a></code> 
        </dt>
        <dd>Evaluate the decay of correlation versus distance between features</dd>
      </dl><dl><dt>
          
          <code><a href="extractCorrelationScores.html">extractCorrelationScores()</a></code> 
        </dt>
        <dd>Extract sample-level correlation scores</dd>
      </dl><dl><dt>
          
          <code><a href="filterClusters.html">filterClusters()</a></code> 
        </dt>
        <dd>Extract subset of clusters</dd>
      </dl><dl><dt>
          
          <code><a href="getClusterNames.html">getClusterNames()</a></code> 
        </dt>
        <dd>Get name of each cluster</dd>
      </dl><dl><dt>
          
          <code><a href="getClusterRanges.html">getClusterRanges()</a></code> 
        </dt>
        <dd>Get genome coordinates for each cluster</dd>
      </dl><dl><dt>
          
          <code><a href="getFeaturesInCluster.html">getFeaturesInCluster()</a></code> 
        </dt>
        <dd>Get feature names in selected cluster</dd>
      </dl><dl><dt>
          
          <code><a href="getFeaturesInClusterList.html">getFeaturesInClusterList()</a></code> 
        </dt>
        <dd>Get feature names in selected cluster</dd>
      </dl><dl><dt>
          
          <code><a href="getPeakDistances.html">getPeakDistances()</a></code> 
        </dt>
        <dd>Compute distance between peaks</dd>
      </dl><dl><dt>
          
          <code><a href="getSubset-methods.html">getSubset()</a></code> 
        </dt>
        <dd>Extract subset of data points</dd>
      </dl><dl><dt>
          
          <code><a href="get_exon_coords.html">get_exon_coords()</a></code> 
        </dt>
        <dd>Get coordinates of exons</dd>
      </dl><dl><dt>
          
          <code><a href="ggplot_by_sampling.html">ggplot_by_sampling()</a></code> 
        </dt>
        <dd>Plot by subsampling in each bin</dd>
      </dl><dl><dt>
          
          <code><a href="jaccard.html">jaccard()</a></code> 
        </dt>
        <dd>Evaluate Jaccard index</dd>
      </dl><dl><dt>
          
          <code><a href="makeImageRect.html">makeImageRect()</a></code> 
        </dt>
        <dd>Convert correlation matrix into triangle plot</dd>
      </dl><dl><dt>
          
          <code><a href="plotClusterSegments.html">plotClusterSegments()</a></code> 
        </dt>
        <dd>Plot cluster segments</dd>
      </dl><dl><dt>
          
          <code><a href="plotCompareCorr.html">plotCompareCorr()</a></code> 
        </dt>
        <dd>Plot two correlation matrices together</dd>
      </dl><dl><dt>
          
          <code><a href="plotCorrDecay.html">plotCorrDecay()</a></code> 
        </dt>
        <dd>Plot correlation delay</dd>
      </dl><dl><dt>
          
          <code><a href="plotCorrTriangle.html">plotCorrTriangle()</a></code> 
        </dt>
        <dd>Plot triangle of correlation matrix</dd>
      </dl><dl><dt>
          
          <code><a href="plotDecorate.html">plotDecorate()</a></code> 
        </dt>
        <dd>Plot decorate analysis</dd>
      </dl><dl><dt>
          
          <code><a href="plotDensityPoints.html">plotDensityPoints()</a></code> 
        </dt>
        <dd>Plot density as color, add outlier points</dd>
      </dl><dl><dt>
          
          <code><a href="plotEnsGenes.html">plotEnsGenes()</a></code> 
        </dt>
        <dd>Plot ENSEMBL genes</dd>
      </dl><dl><dt>
          
          <code><a href="plotGenes.html">plotGenes()</a></code> 
        </dt>
        <dd>Plot genes from a specified region of the human genome.</dd>
      </dl><dl><dt>
          
          <code><a href="plotPairwiseScatter.html">plotPairwiseScatter()</a></code> 
        </dt>
        <dd>Scatter plot of all pairs of variables stratified by test variable</dd>
      </dl><dl><dt>
          
          <code><a href="plotScatterPairs.html">plotScatterPairs()</a></code> 
        </dt>
        <dd>Scatter plot of all pairs of variables stratified by test variable</dd>
      </dl><dl><dt>
          
          <code><a href="retainClusters.html">retainClusters()</a></code> 
        </dt>
        <dd>Retain clusters by applying filter</dd>
      </dl><dl><dt>
          
          <code><a href="runFastStat.html">runFastStat()</a></code> 
        </dt>
        <dd>Test difference in correlation using closed form tests</dd>
      </dl><dl><dt>
          
          <code><a href="runOrderedClustering.html">runOrderedClustering()</a></code> 
        </dt>
        <dd>Run hierarchical clustering preserving order</dd>
      </dl><dl><dt>
          
          <code><a href="runOrderedClusteringGenome.html">runOrderedClusteringGenome()</a></code> 
        </dt>
        <dd>Run hierarchical clustering preserving order</dd>
      </dl><dl><dt>
          
          <code><a href="runPermutedData.html">runPermutedData()</a></code> 
        </dt>
        <dd>Run hierarchical clustering permuting features</dd>
      </dl><dl><dt>
          
          <code><a href="sLEDresults-class.html">sLEDresults-class</a></code> 
        </dt>
        <dd>An S4 class that stores results of sLED analysis</dd>
      </dl><dl><dt>
          
          <code><a href="scoreClusters.html">scoreClusters()</a></code> 
        </dt>
        <dd>Compute scores for each cluster</dd>
      </dl><dl><dt>
          
          <code><a href="sle.score.html">sle.score()</a></code> 
        </dt>
        <dd>Score impact of each sample on sparse leading eigen-value</dd>
      </dl><dl><dt>
          
          <code><a href="sle.test.html">sle.test()</a></code> 
        </dt>
        <dd>Test association between sparse leading eigen-value and variable</dd>
      </dl><dl><dt>
          
          <code><a href="sub-epiclustDiscreteListContain-ANY-ANY-ANY-method.html">`[`(<i>&lt;epiclustDiscreteListContain&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> 
        </dt>
        <dd>Allow subsetting of epiclustDiscreteListContain</dd>
      </dl><dl><dt>
          
          <code><a href="summary-sLEDresults-method.html">summary(<i>&lt;sLEDresults&gt;</i>)</a></code> 
        </dt>
        <dd>Summarize sLED analysis</dd>
      </dl><dl><dt>
          
          <code><a href="whichCluster.html">whichCluster()</a></code> 
        </dt>
        <dd>Find which cluster a peak is in</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

